Split internal distribution handling into separate internal plugin

breskeby opened the pull request:
This PR was originally named "Extract distribution definition from dist-download plugin" but changed after the discussions and feedback I got on this below.

This PR extracts the internal distribution resolving logic out of the distribution download plugin. In general I think that for plugins used by external builds we should aim to not mix their logic with additional requirements of our internal builds and instead follow that pattern potentially with other plugins too.
One potential follow up might be to not even package internal only plugins and logic with build tools. 

The motivation behind this is:

1. Makes plugins used by external builds easier
2. Doesn't mix internal logic only into general purpose plugins.
3. This makes testing different concerns of distribution resolution simpler.
4. Make further improvements in handling distributions easier. I think there's potential for further improvements of the distribution handling (e.g. see https://github.com/elastic/elasticsearch/pull/57730) or revisit the distribution type per project approach as the reason this was introduced (see https://github.com/elastic/elasticsearch/blob/2899e03217a264d9819d3d46bd7e9c82a4b2a0bf/distribution/packages/build.gradle#L429-L430) seems not valid anymore as dependency substitution is not in use anymore.

breskeby commented on buildSrc/src/integTest/java/org/elasticsearch/gradle/DistributionDownloadPluginIT.java:
```diff
@@ -164,14 +164,6 @@ private void checkService(
         }
     }
 
-    private void assertFileDistro(String version, String type, String platform, String flavor, Boolean bundledJdk, String... sysProps)
```
not used before this PR already

rjernst commented:
@breskeby Can you explain more what this split actually solves? To me it splits up code that belongs together. It is true we can rethink the project-per-distribution as we don't use substitutions anymore, but I don't think this split is necessary for such a refactoring.

breskeby commented:
I think there's value in separating the logic for 1) declaring distributions and validating them (what's currently in the finalizeValues() logic) from 2) the complex logic of how those distributions are resolved or build. 
Related to  #57730 I'd ultimately like to have two different plugins for building distributions locally and resolving them remotely. They'll share a common contract to be transient for elasticsearch build so the end user shouldn't care how a version is built, but we could have a the remote resolve only plugin being used by external plugins etc without carrying boilerplate they're never using and for example get rid of some more "isExternal" if else constructs in our plugins. 
All this requires a common base understanding of what a distribution is, independent from how it's resolved. To some degree similar to how dependencies are handled: You define them in your build and you use them at certain places like compiling or running tests, but the underlying mechanism on how they're resolved is defined in a separate concept, the repository definitions that can be different for different users. 

At the moment the distribution-download plugin relies on an implicit contract with the build by requiring certain tasks to be defined by the plugin user (the archive tasks that live in :distribution:archives) that should be defined in the plugin itself. That logic would also move into the build local plugin and expose a common api (e.g. a copy spec) that allows the user to declare what the distribution consists of. Removing these implicit wiring between plugins and build scripts and make the build scripts smaller and less complex and will ultimately also help us when moving to the Gradle Kotlin DSL which is with the current amount of custom script logic in our build not practicable at all IMO. 

Furthermore the distribution-download plugin is also tangled with many other plugins (Rest testing, Test Clusters, ). I haven't looked through all of this yet but I'm sure we can use more focussed smaller plugins to detangle some of the wiring between those plugins by e.g. having some kind of api layer for distributions in one layer and make it clearer of what functionality of a plugin is actually used.

From the end user perspective that PR doesn't solve anything yet, but I think provides a step forward to looking into what I described above. The change is intentionally small as I prefer a set of smaller steps instead of big bang changes, but I really really think this is a good path to go. I'm happy to discuss and also happy to have your input on what you maybe consider higher priority in our build.

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Build)

rjernst commented:
Thanks for the explanation. I appreciate the goal, and see this work long term as a good thing, both for our sanity in the build, and external plugin developers. My hesitation is with "base" plugins in general. The two implementations you describe actually have significant overlap. In fact, I would posit the internal plugin for our build is a superset of the functionality for external developers. When building internally, we still pull distributions from artifacts.elastic.co (in bwc tests), we just also have locally built snapshot versions for most of the testing. So I think this could be done by have the DistributionDownloadPlugin be what external plugin developers use, and it having appropriate hooks so we can subclass for an InternalDistributionDownloadPlugin which extends the functionality to allow our locally built snapshot versions instead of pulling from snapshots.elastic.co (the plugin doesn't currently do this last part, but the intention was to add that functionality eventually so plugin developers could test against unreleased versions).

mark-vieira commented:
I think you are both basically describing the same thing, albeit slight different plugin extension hierarchies. In the end here's what have:

1. A way to describe a distribution you require (aka the "spec"), in our case this is what `ElasticsearchDistribution` is.
2. A way to resolve that spec from a released ES version which we download.
3. A way to resolve that spec from an unreleased ES version which we build from source.

(1) is common to both (2) and (3). External users require only (2) where as we internally need the functionality of (2) and (3). We have two options to compose this.

First, as Rene describes, put (1) in a base plugin, and implement (2) and (3) in separate plugins which are applies as appropriate based on the need.

As Ryan describes (I believe), place (1) and (2) in a plugin, and (3) in a separate one for only internal use.

I don't really have strong feelings on either one. I recommend pushing forward and doing what works out best as we starting pulling things apart. If we find multiple plugins to be awkward, we can easily "merge" them together.

I will note though, we should try and nail down how we are going to handle this internal vs external stuff soon as that's likely to impact what we do here. If and how we share code between `buildSrc` and a purely external build-tools project might come into play here. Initially I wanted to start a brand new project from scratch and share _nothing_ from `buildSrc` but after talking with Ryan we might be able to split out a `common` module in `buildSrc` which we include elsewhere. That said, I'm still apt to share as little as possible if the duplication of code is minimal. By that logic perhaps the "base" plugin model is more appropriate since then we could implement the "external" plugin on that in the external project, and implement our own internally that we can break/change/etc more freely. The core DSL elements could remain shared.

Let's set aside some time to chat about this sometime this week.

breskeby commented:
One use case for illustrating the benefit of having the API elements for `Distribution` in a base plugin could be that an open source user might want to have a different logic for resolving a distribution (e.g. maybe because he has a patched or forked distribution somewhere or wants to pack it differently).  

Regaring splitting download vs install logic it seems like a choice of preference how to structure the split. I personally much prefer plugin compositions instead of subclassing existing plugin classes. Plugin composition feels more natural to me as this is the approach Gradle itself uses to split capabilities from conventions. It's also more flexible when composing multiple functionalities in the long run instead of subclassing. Furthermore it makes tracing hotspots in our build logic via the gradle scans easier I think.
So instead of having a subclass of a `DownloadDistributionPlugin`, the `InternalDistributionPlugin` would apply `DownloadDistributionPlugin` internally plus add functionality to build local distributions. 

@mark-vieira yeah lets find the time to chat about the external vs. internal _problem_ There's a lot I don't have the full picture of I guess.

breskeby committed:
Split internal distribution logic into separate plugin

breskeby committed:
Remove moved func test from DistributionDownloadPluginIT

breskeby committed:
Ensure ordering is reproducible for registered distribution resolutions

breskeby committed:
Fix ordering of dist resolution regisrations

breskeby committed:
Some cleanup

breskeby committed:
Assert internal plugin is only appied on internal build

breskeby commented:
@elasticmachine test this please

breskeby commented:
@elasticmachine test this please

breskeby committed:
Polishing

- extract common abstract test specification for plugin func tests
- update javadocs

breskeby commented on buildSrc/src/main/java/org/elasticsearch/gradle/test/DistroTestPlugin.java:
```diff
@@ -89,7 +90,11 @@
     public void apply(Project project) {
         project.getRootProject().getPluginManager().apply(DockerSupportPlugin.class);
-        project.getPluginManager().apply(DistributionDownloadPlugin.class);
+        if (BuildParams.isInternal()) {
```
We kicked the can of differentiating between internal and external builds further to here for now. 
Is this plugin even considered to be used externally. Otherwise we can remove this differentiation here and just apply the internal plugin

mark-vieira replied:
I believe it is since it's needed for integration testing plugin projects.

mark-vieira replied:
I was referring to `DistributionDownloadPlugin` which is the plugin that's affected by this line change, not `DistroTestPlugin`, which is simply utilizing the download plugin.

rjernst replied:
I thought the original question was whether we needed to have internal/external behavior here, in DistroTestPlugin?

mark-vieira replied:
Oh. In that case, no. But in `DistroDownloadPlugin`, yes. Sorry, I got confused.

breskeby replied:
thank you guys for clarification.

breskeby commented on buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java:
```diff
@@ -49,7 +52,13 @@
     public void apply(Project project) {
-        project.getPlugins().apply(DistributionDownloadPlugin.class);
+        project.getRootProject().getPluginManager().apply(GlobalBuildInfoPlugin.class);
+        if (BuildParams.isInternal()) {
```
We kicked the can of differentiating between internal and external builds further to here for now. We might want to apply similar internal / external plugin distinction later for that plugin too

breskeby committed:
DistroTestPlugin is used internal only

breskeby commented:
@elasticmachine test this please

breskeby commented:
@rjernst & @mark-vieira can you please review this (and tick it off that you reviewed it :))

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalDistributionDownloadPlugin.java:
```diff
@@ -0,0 +1,150 @@
+        resolutions.register("bwb", distributionResolution -> distributionResolution.setResolver((project, distribution) -> {
+            BwcVersions.UnreleasedVersionInfo unreleasedInfo = bwcVersions.unreleasedInfo(Version.fromString(distribution.getVersion()));
+            if (unreleasedInfo != null) {
+                assert distribution.getBundledJdk();
```
Should we add a message here so that if folks incorrectly configure a snapshot bwc distribution w/o bundled JDK that we return an error as such?

breskeby replied:
I changed this to throw a proper GradleException with an error message.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java:
```diff
@@ -45,71 +41,83 @@
  */
 public class DistributionDownloadPlugin implements Plugin<Project> {
 
+    static final String RESOLUTION_CONTAINER_NAME = "elasticsearch_distributions-resolutions";
```
Or perhaps it doesn't really matter since we don't really expect folks to configure this in build scripts?

breskeby replied:
I fixed the name to not be mixed. Working it into the `elasticsearch_distributions` extensions is more work as this is just a plain named container. We don't even use the name in the plugins and I don't expect anyone to configure this in a script.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalDistributionDownloadPlugin.java:
```diff
@@ -0,0 +1,150 @@
+        // this is needed for isInternal
+        project.getRootProject().getPluginManager().apply(GlobalBuildInfoPlugin.class);
+        if (!BuildParams.isInternal()) {
+            throw new GradleException(
```
Nice. At least for now until we physically separate this code such that we don't publish internal stuff this seems like a good pattern to ensure that folks don't "accidentally" leverage internal plugins.

mark-vieira reviewed (approved):
Changes LGTM. I think this is a significant improvement and likely is going to reduce the possibility of us "breaking" external folks by striclty separating this functionality. Now the distro download plugin has nothing to do with resolving internal projects. We're also in a much better position to refactor our internal stuff out of the artifact we publish externally. Thanks, Rene.

breskeby committed:
Apply review feedback

- rename resolve strategy container
- fix naming of bwc resolving strategy
- provide proper failure if unexpected non jdk bundled bwc version is requested

breskeby commented on buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalDistributionDownloadPlugin.java:
```diff
@@ -0,0 +1,150 @@
+            return null;
+        }));
+
+        resolutions.register("bwb", distributionResolution -> distributionResolution.setResolver((project, distribution) -> {
```
fixed
